x/tally/keeper/gas_meter.go

```diff
@@ -0,0 +1,190 @@
+
+	// Append distribution messages for data proxies.
+	for _, proxy := range gasMeter.Proxies {
+		proxyDist := types.NewDataProxyReward(proxy.PublicKey, proxy.Amount, gasMeter.GasPrice())
+		dists = append(dists, proxyDist)
+		attrs = append(attrs, sdk.NewAttribute(types.AttributeDataProxyGas,
+			fmt.Sprintf("%s,%s", hex.EncodeToString(proxy.PublicKey), proxy.Amount.String())))
+	}
+
+	// Append distribution messages for executors, burning a portion of their
+	// payouts in case of a reduced payout scenario.
+	reducedPayoutBurn := math.ZeroInt()
+	for _, executor := range gasMeter.Executors {
+		payoutAmt := executor.Amount
+		if gasMeter.ReducedPayout {
+			burnAmt := burnRatio.MulInt(executor.Amount).TruncateInt()
+			payoutAmt = executor.Amount.Sub(burnAmt)
+			reducedPayoutBurn = reducedPayoutBurn.Add(burnAmt.Mul(gasMeter.GasPrice()))
+		}
+
+		executorDist := types.NewExecutorReward(executor.PublicKey, payoutAmt, gasMeter.GasPrice())
+		dists = append(dists, executorDist)
+		attrs = append(attrs, sdk.NewAttribute(types.AttributeExecutorGas,
+			fmt.Sprintf("%s,%s", executor.PublicKey, payoutAmt.String())))
+	}
+
+	dists[0].Burn.Amount = dists[0].Burn.Amount.Add(reducedPayoutBurn.Mul(gasMeter.GasPrice()))
+	attrs = append(attrs, sdk.NewAttribute(types.AttributeReducedPayoutBurn, reducedPayoutBurn.String()))
+
+	ctx.EventManager().EmitEvent(sdk.NewEvent(types.EventTypeGasMeter, attrs...))
+
+	return dists
+}
+
+// MeterProxyGas computes and records the gas consumption of data proxies given
+// proxy public keys in basic consensus and the request's replication factor.
+func (k Keeper) MeterProxyGas(ctx sdk.Context, proxyPubKeys []string, replicationFactor uint16, gasMeter *types.GasMeter) {
+	if len(proxyPubKeys) == 0 || gasMeter.RemainingExecGas() == 0 {
+		return
+	}
+
+	for _, pubKey := range proxyPubKeys {
+		pubKeyBytes, err := hex.DecodeString(pubKey)
+		if err != nil {
+			k.Logger(ctx).Error("failed to decode proxy public key", "error", err, "public_key", pubKey)
+			continue
+		}
+		proxyConfig, err := k.dataProxyKeeper.GetDataProxyConfig(ctx, pubKeyBytes)
+		if err != nil {
+			k.Logger(ctx).Error("failed to get proxy config", "error", err, "public_key", pubKey)
+			continue
+		}
+
+		gasUsedPerExec := proxyConfig.Fee.Amount.Quo(gasMeter.GasPrice()).Uint64()
+		gasUsedPerExec = min(gasUsedPerExec, gasMeter.RemainingExecGas()/uint64(replicationFactor))
+		gasUsed := gasUsedPerExec * uint64(replicationFactor)
+
+		gasMeter.ConsumeExecGasForProxy(pubKeyBytes, gasUsed)
+	}
+}
+
+// MeterExecutorGasFallback computes and records the gas consumption of committers
+// of a data request when basic consensus has not been reached. If checkReveal is
+// set to true, it will only consume gas for committers that have also revealed.
+func MeterExecutorGasFallback(req types.Request, gasCostCommit uint64, gasMeter *types.GasMeter, checkReveal bool) {
```

Comment:
I think we can infer checkReveal. I can't think of a reason where we would want to reward a fallback to an executor who could have revealed but didn't.